### PR TITLE
 	Prototype of mvc services using DependencyContext

### PR DESCRIPTION
**Pull request by pakrym**
HelloMvc using https://github.com/dotnet/cli/pull/634
@davidfowl

### CODE REVIEW

**Comment by pakrym** [comment]
1. Does not have compilation settings yet.
2. I'd prefer having some transition from DependencyContext.CompilationLibraries to refs dll file in API

File: Microsoft.AspNet.Mvc.Prototypes/project.json [hunk 1]
```diff
@@ -0,0 +1,14 @@
+{
+  "dependencies": {
+    "Microsoft.AspNet.Mvc.Core": "6.0.0-*",
+    "Microsoft.Extensions.CompilationAbstractions": "1.0.0-rc2-*",
+    "Microsoft.Extensions.DependencyModel": "1.0.0-*",
+    "Microsoft.CodeAnalysis": "1.1.0-*"
+  },
+  "frameworks": {
+    "dnx451": { },
+    "dnxcore50": {
```
Comment (davidfowl) [inline Microsoft.AspNet.Mvc.Prototypes/project.json]: This dependency should not be in here. It should only ever exist in applications not libraries

Comment (pakrym) [inline Microsoft.AspNet.Mvc.Prototypes/project.json]: Which one?

**Comment by pakrym** [comment]
Do we want to add Try looking in %DOTNET_PACKAGES% folder first?

File: Microsoft.AspNet.Mvc.Prototypes/project.json (hunk 1 above)
Comment (pakrym) [inline Microsoft.AspNet.Mvc.Prototypes/project.json]: Ah, got it

Comment (pakrym) [inline Microsoft.AspNet.Mvc.Prototypes/project.json]: Fixed

File: Microsoft.AspNet.Mvc.Prototypes/DependencyContextReferenceResolver.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+namespace Microsoft.AspNet.Mvc
+{
+    public class DependencyContextReferenceResolver
+    {
+        private Lazy<IEnumerable<MetadataReference>> _references = new Lazy<IEnumerable<MetadataReference>>(GetReferences);
+
+        private static IEnumerable<MetadataReference> GetReferences()
+        {
+            var entryAssembly = (Assembly)typeof(Assembly).GetTypeInfo().GetDeclaredMethod("GetEntryAssembly").Invoke(null, null);
```
Comment (davidfowl) [inline Microsoft.AspNet.Mvc.Prototypes/DependencyContextReferenceResolver.cs]: is does this end up being non null?

File: Microsoft.AspNet.Mvc.Prototypes/DependencyContextAssemblyProvider.cs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+            foreach (var library in dependsOnMvc)
+            {
+                if (library.PackageName.StartsWith(MicrosoftAspnetMvcPrefix))
+                {
+                    // skip mvc libs
+                    continue;
+                }
+
+                if (library.LibraryType == "project")
```
Comment (davidfowl) [inline Microsoft.AspNet.Mvc.Prototypes/DependencyContextAssemblyProvider.cs]: You shouldn't have to check the type. Assemblies should have the right thing in either case

Comment (pakrym) [inline Microsoft.AspNet.Mvc.Prototypes/DependencyContextAssemblyProvider.cs]: project.lock.json does't have assembly name for "project" libraries, i cant add it to dependency context generation though

File: Microsoft.AspNet.Mvc.Prototypes/DependencyContextReferenceResolver.cs (hunk 2 above)
Comment (pakrym) [inline Microsoft.AspNet.Mvc.Prototypes/DependencyContextReferenceResolver.cs]: Yep. Also I'd like to put this code in some other place, maybe add class to DependencyModel which would be able to return Assembly for runtime library and location for compile time library, would be more user friendly this way.

File: Microsoft.AspNet.Mvc.Prototypes/DependencyContextAssemblyProvider.cs (hunk 3 above)
Comment (davidfowl) [inline Microsoft.AspNet.Mvc.Prototypes/DependencyContextAssemblyProvider.cs]: Yes, the dependency context should normalize the data. You don't really have projects at runtime so the entire concept should go away.

File: Microsoft.AspNet.Mvc.Prototypes/project.json [hunk 4]
```diff
@@ -0,0 +1,17 @@
+{
+  "dependencies": {
+    "Microsoft.AspNet.Mvc.Core": "6.0.0-*",
+    "Microsoft.Extensions.CompilationAbstractions": "1.0.0-rc2-*",
+    "Microsoft.Extensions.DependencyModel": "1.0.0-*",
+    "Microsoft.CodeAnalysis": "1.1.0-*"
+  },
+  "frameworks": {
```
Comment (davidfowl) [inline Microsoft.AspNet.Mvc.Prototypes/project.json]: These are also the wrong frameworks. Should be dotnet5.4 and net451

File: HelloMvc/Startup.cs [hunk 5]
```diff
@@ -21,15 +17,12 @@ public Startup(IApplicationEnvironment applicationEnvironment)
...
             _applicationEnvironment = applicationEnvironment;
         }
-        
+
         public void ConfigureServices(IServiceCollection services)
         {
-            // We need to disable register an empty exporter so that we can add references
-            // This will go away eventually see
-            services.AddSingleton<ILibraryExporter, NullExporter>();
+            var refs = new DependencyContextReferenceResolver();
```
Comment (davidfowl) [inline HelloMvc/Startup.cs]: Try writing something like this `https://github.com/aspnet/Mvc/blob/c5b6efd6bf1d488992c6c6adc262b56200029e21/src/Microsoft.AspNet.Mvc.Razor/RazorViewEngineOptionsSetup.cs#L17`

That way we hide all of this stuff away into the 1 liner.

File: HelloMvc/Startup.cs [hunk 6]
```diff
@@ -21,15 +17,12 @@ public Startup(IApplicationEnvironment applicationEnvironment)
...
+
 
-            // ILibraryManager isn't available yet so we need to explicitly add the assemblies
-            // that we want to find controllers in. 
             var assemblyProvider = new StaticAssemblyProvider();
```
Comment (davidfowl) [inline HelloMvc/Startup.cs]: This should be gone as well since you have the `DependencyContextAssemblyProvider`

File: Microsoft.AspNet.Mvc.Prototypes/DependencyContextAssemblyProvider.cs [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    internal class DependencyContextAssemblyProvider: IAssemblyProvider
+    {
+        private const string MicrosoftAspnetMvcPrefix = "Microsoft.AspNet.Mvc";
+
+        private Lazy<IEnumerable<Assembly>> _candidateAssemblies = new Lazy<IEnumerable<Assembly>>(GetCandidateAssemblies);
+
+        private static IEnumerable<Assembly> GetCandidateAssemblies()
+        {
+            var dependencyContext = DependencyContext.Load();
+            var dependsOnMvc = dependencyContext.RuntimeLibraries.Where(DependsOnMvc).ToArray();
```
Comment (davidfowl) [inline Microsoft.AspNet.Mvc.Prototypes/DependencyContextAssemblyProvider.cs]: Why ToArray()? you only enumerate once

**Comment by davidfowl** [comment]
@pakrym Needs a rebase

**Comment by pakrym** [comment]
:up: :date: Cleaned up and added compilation options based on https://github.com/dotnet/cli/pull/634

**Comment by davidfowl** [comment]
@pakrym conflicts

File: HelloMvc/project.json [hunk 8]
```diff
@@ -1,16 +1,19 @@
...
+      "Microsoft.NETCore.Platforms": "1.0.1-beta-*",
+      "Microsoft.AspNet.Mvc": "6.0.0-*",
+      "Microsoft.AspNet.Diagnostics": "1.0.0-*",
+      "Microsoft.AspNet.Server.Kestrel": "1.0.0-*",
+      "System.Runtime.Serialization.Primitives": "4.1.0-*",
     },
     "frameworks": {
-        "dnxcore50": { }
+        "net451": { },
+        "dotnet5.5": { }
```
Comment (davidfowl) [inline HelloMvc/project.json]: This is wrong too. Don't change the target frameworks or add any. The CoreCLR and full samples are split for now.

**Comment by davidfowl** [comment]
This needs to be done over based on the latest stuff:
- Move https://github.com/davidfowl/dotnetcli-aspnet5/tree/master/HelloMvc/Infrastructure to the Prototypes project
- Make your changes in there.

File: Microsoft.AspNet.Mvc.Prototypes/DependencyContextReferenceResolver.cs [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+            options.CompilationCallback = (context) =>
+            {
+                previousCallback?.Invoke(context);
+
+                SetCompilationOptions(context, compilationOptions);
+            };
+        }
+
+        private void SetCompilationOptions(RoslynCompilationContext context, CompilationOptions compilationOptions)
```
Comment (davidfowl) [inline Microsoft.AspNet.Mvc.Prototypes/DependencyContextReferenceResolver.cs]: For razor pages we don't care about most of these settings actually:

- The diagnostics can be hard coded into MVC itself, they shouldn't be in here (i.e. WithSpecificDiagnosticOptions).
- Platform should be ignored
- optimize is tricky, for runtime compilation the settings should match runtimey things (like IHostingEnvironment)

File: Microsoft.AspNet.Mvc.Prototypes/DependencyContextReferenceResolver.cs [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+using Microsoft.CodeAnalysis;
+using Microsoft.Extensions.PlatformAbstractions;
+using Microsoft.Extensions.DependencyModel;
+
+using CompilationOptions = Microsoft.Extensions.DependencyModel.CompilationOptions;
+using Platform = Microsoft.CodeAnalysis.Platform;
+
+namespace Microsoft.AspNet.Mvc
+{
+    public class CliRazorViewEngineOptionsSetup: RazorViewEngineOptionsSetup
```
Comment (davidfowl) [inline Microsoft.AspNet.Mvc.Prototypes/DependencyContextReferenceResolver.cs]: Why derive from the existing one? These compose.

File: Microsoft.AspNet.Mvc.Prototypes/DependencyContextAssemblyProvider.cs [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+            {
+                if (library.PackageName.StartsWith(MicrosoftAspnetMvcPrefix))
+                {
+                    // skip mvc libs
+                    continue;
+                }
+
+                foreach (var assembly in library.Assemblies)
+                {
+                    yield return Assembly.Load(new AssemblyName(Path.GetFileNameWithoutExtension(assembly)));
```
Comment (davidfowl) [inline Microsoft.AspNet.Mvc.Prototypes/DependencyContextAssemblyProvider.cs]: These are relative paths right?

Comment (pakrym) [inline Microsoft.AspNet.Mvc.Prototypes/DependencyContextAssemblyProvider.cs]: Yes

File: Microsoft.AspNet.Mvc.Prototypes/DependencyContextReferenceResolver.cs (hunk 9 above)
Comment (pakrym) [inline Microsoft.AspNet.Mvc.Prototypes/DependencyContextReferenceResolver.cs]: We have these setting for free anyway, why don't use them?

**Comment by pakrym** [comment]
@davidfowl rebase

File: Microsoft.AspNet.Mvc.Prototypes/DependencyContextReferenceResolver.cs (hunk 9 above)
Comment (pranavkm) [inline Microsoft.AspNet.Mvc.Prototypes/DependencyContextReferenceResolver.cs]: :+1: on ignoring Platform. It's a bit weird to have views targeting specific platforms.

**Comment by pakrym** [comment]
now?

File: HelloMvc/Infrastructure/DependencyContextAssemblyProvider.cs [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+{
+    internal class DependencyContextAssemblyProvider: IAssemblyProvider
+    {
+        private const string MicrosoftAspnetMvcPrefix = "Microsoft.AspNet.Mvc";
+
+        private Lazy<IEnumerable<Assembly>> _candidateAssemblies = new Lazy<IEnumerable<Assembly>>(GetCandidateAssemblies);
+
+        private static IEnumerable<Assembly> GetCandidateAssemblies()
+        {
+            var dependencyContext = DependencyContext.Load();
```
Comment (davidfowl) [inline HelloMvc/Infrastructure/DependencyContextAssemblyProvider.cs]: Feels like this should be a property.

Comment (pranavkm) [inline HelloMvc/Infrastructure/DependencyContextAssemblyProvider.cs]: `DependencyContext.Instance` which does lazy loading. If there's going to be exactly one instance, the `Load` seems weird.

Comment (pakrym) [inline HelloMvc/Infrastructure/DependencyContextAssemblyProvider.cs]: OK, will do in another pr

**Comment by pakrym** [comment]
Rebased, removed platform fallthrough
